File: UserUtility.m
//
//  UserUtility.m
//  mage-ios-sdk
//
//

@import DateTools;

#import "UserUtility.h"
#import "MageSessionManager.h"
#import "MageServer.h"
#import "StoredPassword.h"

@interface UserUtility()

@property BOOL expired;

@end

@implementation UserUtility

+ (id) singleton {
    static UserUtility *userUtility = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        userUtility = [[self alloc] init];
    });
    return userUtility;
}

- (id) init {
    if (self = [super init]) {
        self.expired = NO;
    }
    return self;
}

- (void) resetExpiration {
    self.expired = NO;
}

- (void) acceptConsent {
    NSUserDefaults *defaults = [NSUserDefaults standardUserDefaults];
    NSMutableDictionary *loginParameters = [[defaults objectForKey:@"loginParameters"] mutableCopy];
    [loginParameters setValue:@"agree" forKey:@"acceptedConsent"];
    [defaults setObject:loginParameters forKey:@"loginParameters"];
    [defaults synchronize];
}

- (BOOL) isTokenExpired{
    if (self.expired) return YES;
    
    NSUserDefaults *defaults = [NSUserDefaults standardUserDefaults];
    NSDictionary *loginParameters = [defaults objectForKey:@"loginParameters"];
    
    NSString *acceptedConsent = [loginParameters objectForKey:@"acceptedConsent"];
    
    NSDate *tokenExpirationDate = [loginParameters objectForKey:@"tokenExpirationDate"];
    if (acceptedConsent != nil && [acceptedConsent isEqualToString:@"agree"] && tokenExpirationDate != nil && [tokenExpirationDate isKindOfClass:NSDate.class]) {
        NSDate *currentDate = [NSDate date];
        NSLog(@"current date %@ token expiration %@", currentDate, tokenExpirationDate);
        self.expired = [currentDate isLaterThan:tokenExpirationDate];
        if (self.expired) {
            [self expireToken];
            [[NSNotificationCenter defaultCenter] postNotificationName:MAGETokenExpiredNotification object:nil];
        }
        return self.expired;
    }
    self.expired = YES;
    return self.expired;
}

- (void) expireToken {
    [StoredPassword clearToken];
	NSUserDefaults *defaults = [NSUserDefaults standardUserDefaults];
    NSMutableDictionary *loginParameters = [[defaults objectForKey:@"loginParameters"] mutableCopy];
    
    [loginParameters removeObjectForKey:@"tokenExpirationDate"];
    [loginParameters removeObjectForKey:@"acceptedConsent"];
    
    [[MageSessionManager manager] clearToken];
    
    [defaults setObject:loginParameters forKey:@"loginParameters"];
    
    [defaults removeObjectForKey:@"loginType"];
    
    [defaults synchronize];
    self.expired = YES;
}

- (void) logoutWithCompletion: (void (^)(void)) completion {
    NSString *url = [NSString stringWithFormat:@"%@/api/logout", [MageServer baseURL]];

    MageSessionManager *manager = [MageSessionManager manager];
    __weak typeof(self) weakSelf = self;

    NSURLSessionDataTask *task = [manager POST_TASK:url parameters:nil progress:nil success:^(NSURLSessionTask *task, id responseObject) {
        NSLog(@"Logged out");
        [weakSelf expireToken];
        completion();
    } failure:^(NSURLSessionTask *operation, NSError *error) {
        NSLog(@"Error: %@", error);
        [weakSelf expireToken];
        completion();
    }];
    [manager addTask:task];
}

@end

File: sdk/UserUtility.swift
//
//  UserUtility.m
//  mage-ios-sdk
//
//

import Foundation

@objc public class UserUtility: NSObject {
    
    var expired: Bool = false
    
    @objc public static let singleton = UserUtility()
    
    private override init() {
        expired = false
    }
    
    @objc public var isTokenExpired: Bool {
        get {
            if expired {
                return true
            }
            let loginParameters = UserDefaults.standard.loginParameters
            
            if let loginParameters = loginParameters,
               let acceptedConsent = loginParameters[LoginParametersKey.acceptedConsent.key] as? String,
               let tokenExpirationDate = loginParameters[LoginParametersKey.tokenExpirationDate.key] as? Date,
               acceptedConsent == "agree" {
                let currentDate = Date()
                NSLog("current date \(currentDate) token expiration \(tokenExpirationDate)")
                expired = currentDate > tokenExpirationDate
                if expired {
                    self.expireToken()
                    NotificationCenter.default.post(name: .MAGETokenExpiredNotification, object: nil)
                }
                return expired
            }
            expired = true
            return expired
        }
    }
    
    @objc public func expireToken() {
        StoredPassword.clearToken()
        var loginParameters = UserDefaults.standard.loginParameters ?? [:]
        loginParameters.removeValue(forKey: LoginParametersKey.tokenExpirationDate.key)
        loginParameters.removeValue(forKey: LoginParametersKey.acceptedConsent.key)
        
        MageSessionManager.shared().clearToken()
        
        UserDefaults.standard.loginParameters = loginParameters
        UserDefaults.standard.loginType = nil
        
        self.expired = true
    }
    
    @objc public func resetExpiration() {
        self.expired = false
    }
    
    @objc public func acceptConsent() {
        var loginParameters = UserDefaults.standard.loginParameters ?? [:]
        loginParameters[LoginParametersKey.acceptedConsent.key] = LoginParametersKey.agree.key
        UserDefaults.standard.loginParameters = loginParameters
    }
    
    @objc public func logout(completion: @escaping () -> Void) {
        guard let baseUrl = MageServer.baseURL() else {
            completion()
            return
        }
        Task { @MainActor in
            await LogoutService.logout(baseURL: baseUrl)
            self.expireToken()
            completion()
        }
    } 
}

extension UserUtility {
    @objc class func appVersionString() -> String {
        let info = Bundle.main.infoDictionary
        let ver = info?["CFBundleShortVersionString"] as? String ?? "0"
        let build = info?["CFBundleVersion"] as? String ?? "0"
        return "\(ver) (\(build))"
    }
}
